File: leetcode/091_Decode_Ways.py
class Solution(object):
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s: 
            return 0
        dp = [1] + [0] * len(s)
        # 如果i 不为0, 那么至少可以是dp[i]次
        # 考虑i 和 i-1 位，如果i>0 并且 i-1 和第 i 位可以组合成10到26之间的数字，那么至少可以dp[i-1]次
        for i in range(len(s)):
            dp[i + 1] = dp[i] * (int(s[i]) > 0) + dp[i - 1] * (i and 10 <= int(s[i - 1: i + 1]) <= 26)
        return dp[-1]
        
        
# # 此程序超时        
# class Solution(object):
#     def numDecodings(self, s):
#         """
#         :type s: str
#         :rtype: int
#         """
#         if len(s) == 0: return 0
#         return self.getnum(s)
        
#     def getnum(self, s):
#         # 如果长度为1 或者为2
#         if len(s) == 1:
#             return 1 if self.test(s) else 0
#         if len(s) == 2:
#             if self.test(s[0]) and self.test(s[1]):
#                 return 2 if self.test(s) else 1
#             return 1 if self.test(s) else 0
#         # 长度大于2 因为可能包含0，考虑四种情况
#         if not self.test(s[0]) and self.test(s[0:2]):
#             return self.getnum(s[2:])
#         elif self.test(s[0]) and not self.test(s[0:2]):
#             return self.getnum(s[1:])
#         elif self.test(s[0]) and self.test(s[0:2]):
#             return self.getnum(s[1:]) + self.getnum(s[2:])
#         else:
#             return 0 

            
#     def test(self, s):
#         # 测试是否能够转化成A-Z
#         if int(s[0]) == 0:
#             return False
#         if int(s) >= 1 and int(s) <= 26:
#             return True
        
        

File: leetcode/091_Decode_Ways.java
// 2017.07.29
// attention for the initiation 
public class Solution {
    public int numDecodings(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        int n = s.length();
        int[] dp = new int[n + 1];
        dp[0] = 1;  // attention 
        dp[1] = s.charAt(0) == '0' ? 0 : 1; // attention 
        for (int i = 1; i < n; i++) {
            // dp[i + 1] 
            int first = Integer.valueOf(s.substring(i, i + 1));
            int last = Integer.valueOf(s.substring(i - 1, i + 1));
            if (first >= 1 && first <= 9) {
                dp[i + 1] += dp[i];
            } 
            if (last >= 10 && last <= 26) {
                dp[i + 1] += dp[i - 1];
            }
        }
        return dp[n];
    }
}


// memorization method, TLE
class Solution {
    public int numDecodings(String s) {
        if (s == null || s.length() == 0)
            return 0;
        int[] mem = new int[s.length()];
        return helper(s, 0, mem);
    }
    private int helper(String s, int idx, int[] mem) {
        if (idx == s.length()) 
            return 1;
        // one char
        int c1 = 0;
        if (s.charAt(idx) != '0') 
            c1 = helper(s, idx + 1, mem);
        // two char
        int c2 = 0;
        if (idx + 1 < s.length() && (s.charAt(idx) == '1' || (s.charAt(idx) == '2'  && s.charAt(idx +1) - '0' <= 6))) 
            c2 = helper(s, idx + 2, mem);
        // mem
        mem[idx] = c1 + c2;
        return mem[idx];
    }
}